src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+                if (self.incoming_header.command == .request) {
+                    self.peer = .{ .client = self.incoming_header.client };
+                } else {
+                    self.message_bus.replica_connections += 1;
+                    self.peer = .{ .replica = self.incoming_header.replica };
+                }
+            }
+
+            self.incoming_message = self.message_bus.create_message(self.incoming_header.size) catch unreachable;
+            self.incoming_message.references += 1;
+            self.incoming_message.header.* = self.incoming_header;
+            self.recv_body();
+        }
+
+        fn recv_body(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_body,
+                &self.completion,
+                self.incoming_message.buffer[self.bytes_received..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_recv_body(self: *Connection, completion: *IO.Completion, result: RecvError!usize) void {
+            self.bytes_received += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error receiving data from: {}", .{ peer, err });
+                self.close();
+                return;
+            };
+
+            if (self.bytes_received < self.incoming_message.buffer.len) {
+                // The body has not yet been fully received.
+                self.recv_body();
+                return;
+            }
+            assert(self.bytes_received == self.incoming_message.buffer.len);
+
+            const data = self.incoming_message.buffer[@sizeOf(Header)..self.incoming_header.size];
+            if (self.incoming_header.valid_checksum_data(data)) {
+                self.message_bus.server.on_message(self.incoming_message);
+            } else {
+                log.err("invalid checksum on data received from {}", .{self.peer});
+                self.close();
+            }
+
+            self.incoming_message.references -= 1;
+            self.message_bus.gc(self.incoming_message);
+
+            // Reset state and try to receive the next message.
+            self.incoming_header = undefined;
+            self.incoming_message = undefined;
+            self.bytes_received = 0;
+            self.recv_header();
+        }
+
+        fn send_message(self: *Connection, envelope: *Envelope) void {
+            assert(self.peer == .client);
+            const queue_was_empty = self.send_queue.out == null;
+            self.send_queue.push(envelope);
+            // If the queue was not empty, the message will be sent after the
+            // messages currently being sent.
+            if (queue_was_empty) self.send();
+        }
+
+        fn send(self: *ReplicaConnection) void {
```

Comment:
The context here should be `*Connection` instead of `*ReplicaConnection`? Vive la copypasta! 😄

Reply:
Indeed, I have certainly made my fair share of copy paste mistakes in these commits, not having the compiler to tell me I'm wrong doesn't help things :D